Breadcrumbs.py

```diff
@@ -1,94 +1,107 @@
 # -*- coding: utf-8 -*-
-import sublime, sublime_plugin
-import math
+import re
+import html
+
+import sublime
+import sublime_plugin
 
 try:
   xrange
 except NameError:
   xrange = range
 
+
 def get_tab_size(view):
   return int(view.settings().get('tab_size', 8))
 
-def get_row_indentation(points, view, tab_size, limit=1e20):
-  pos = 0
-  for pt in points:
-    ch = view.substr(pt)
 
+def get_row_indentation(line_start, view, tab_size, indentation_limit=None):
+  if indentation_limit is None:
+    row = view.rowcol(line_start)[0]
+    indentation_limit = view.text_point(row + 1, 0) - line_start
+
+  pos = 0
+  for ch in view.substr(sublime.Region(line_start, line_start + indentation_limit)):
     if ch == '\t':
       pos += tab_size - (pos % tab_size)
 
-    elif ch.isspace():
+    elif ch == ' ':
       pos += 1
 
     else:
       break
 
-    if limit <= pos:
+    if indentation_limit <= pos:
       break
 
   return pos
 
+
 def is_white_row(view, points):
   return all(view.substr(pt).isspace() for pt in reversed(points))
 
-def get_breadcrumb(view, points, limit):
-  for pt in points:
-    ch = view.substr(pt)
-    if not ch.isspace():
-      return view.substr(sublime.Region(pt, min(view.line(pt).b, pt + limit))).strip()
-  return ''
 
-class BreadcrumbsCommand(sublime_plugin.EventListener):
-  def on_selection_modified(self, view):
-    tab_size = get_tab_size(view)
-    settings = sublime.load_settings('Breadcrumbs.sublime-settings')
-    breadcrumb_length_limit = settings.get('breadcrumb_length_limit', 100)
-    separator = settings.get('breadcrumbs_separator', u' ■ ')
-    total_breadcrumbs_length_limit = settings.get('total_breadcrumbs_length_limit', 200)
-    view.erase_status('breadcrumbs')
+def get_breadcrumb(view, points, regex, limit):
+  linestring = view.substr(sublime.Region(min(points), min(view.line(min(points)).b, min(points) + limit)))
+  match = re.search(regex, linestring)
+  if match:
+    return(match.group('name'))
+  else:
+    return None
 
-    if len(view.sel()) == 0:
-      return
 
-    current_row = view.rowcol(view.sel()[0].b)[0]
+def make_breadcrumbs(view, current_row, shorten=False):
+  if len(view.sel()) == 0:
+    return
 
-    def get_row_start(row):
-      return view.text_point(row,0)
+  settings = sublime.load_settings('Breadcrumbs.sublime-settings')
+  tab_size = get_tab_size(view)
+  default_breadcrumb_regex = settings.get('breadcrumb_regex', u'(?P<name>.*)')
+  breadcrumb_regex = view.settings().get('breadcrumb_regex', default_breadcrumb_regex)
+  separator = settings.get('breadcrumbs_separator', u' ')
```

Comment:
> I'm not sure what is the convention for defaults. 

Me neither, I'm basically bailing out of further researching why the regex breaks highlighting. But then I thought keeping the settings and defaults in sync doesn't do anything since we can rely on the settings always being there. Users shouldn't edit the package settings unless they're hacking it (and they can't even reach it properly if installed through package control).
I think the defaults are mostly a good way to document the kind of value you're expecting, and having a fallback so that it doesn't crash if the settings file isn't there doesn't seem like a bad idea. So I reduced the defaults to something that will work and we can leave them be if we ever change any of the default settings.

The best way to present the settings and customisation settings is to show the default settings side by side with the user settings via a menu entry. The application settings do this and packages can do so to but it takes a little effort. It's basically as you describe, and we can remove the table from the README.
Gitgutter does this for instance, I'll find out if we can borrow some of it.